### ISSUE TITLE
[Rule Request]: singleton_private_init

### ISSUE THREAD

**Issue by sindresorhus** [realm/SwiftLint#4774]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request

To create a singleton, you would write something like this:

```swift
final class Foo {
	static let shared = Foo()

	init() {
		// …
	}
}
```

The problem is that it's easy to forget to make the `init()` private, which means you (or someone on your team) could accidentally create instances of it. That happened to me a few times.

In cases it's desired to have both a singleton and instances, the rule can simply be disabled for such cases. I don't think that's common though.

> 1. Why should this rule be added? Share links to existing discussion about what
   the community thinks about this.

It prevents real-world mistakes.

> 2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

**Would trigger**

```swift
final class Foo {
	static let shared = Foo()

	init() {
		// …
	}
}
```

**Would not trigger**

```swift
final class Foo {
	static let shared = Foo()

	private init() {
		// …
	}
}
```

> 4. Should the rule be configurable, if so what parameters should be configurable?

No

> 5. Should the rule be opt-in or enabled by default? Why?
   See [README.md](../README.md#opt-in-rules) for guidelines on when to mark a rule as opt-in.

I guess opt-in.

**Comment by SimplyDanny** [comment]
What if the static field is called `foo` or `default`? Is the name of the instance relevant? What if there are more than the single static instance? Would the rule trigger? More examples might be required to assess this proposal.

**Comment by sindresorhus** [comment]
`.shared` is the most common convention for singletons. I suggest starting simple and only support `.shared` and a single field. Support for more names and configuration can be added at a later time based on demand.

So to make it clear: The rule would only trigger if the class has a static property named `shared` that initiates the class.

**Comment by SimplyDanny** [comment]
Just to mention, I've also seen different spellings of `default`, `instance` and `shared`.
